spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/PredicateOps.scala

```diff
@@ -0,0 +1,93 @@
+package org.apache.spark.sql.cassandra
+
+import org.apache.spark.sql.catalyst.expressions
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.sources
+import org.apache.spark.sql.sources.Filter
+
+/** A unified API for predicates, used by [[PredicatePushDown]].
+  *
+  * Keeps all the Spark-specific stuff out of `PredicatePushDown`
+  * It is also easy to plug-in custom predicate implementations for unit-testing.
+  */
+trait PredicateOps[Predicate] {
+
+  /** Returns the column name of the only column in the predicate.
+    * Will throw `IllegalArgumentException` if predicate is not a single-column predicate. */
```

Comment:
*Nulls are never permitted* in Scala, except two cases:
1. interfacing with Java code; nulls should be immediately de-nullified by wrapping them into Options.
2. critical performance path, where overhead of an Option might be too large and we know the compiler won't elide it; but then they should be very limited in scope - possibly only method local, or class-private. And it should be explicitly documented why nulls were needed.